drivers/vdpa/virtio/virtio_vdpa.c

```diff
@@ -0,0 +1,970 @@
+#include <rte_vfio.h>
+#include <rte_vhost.h>
+#include <rte_vdpa.h>
+#include <vdpa_driver.h>
+#include <rte_kvargs.h>
+
+#include <virtio_api.h>
+#include "virtio_vdpa.h"
+
+RTE_LOG_REGISTER(virtio_vdpa_logtype, pmd.vdpa.virtio, NOTICE);
+#define DRV_LOG(level, fmt, args...) \
+	rte_log(RTE_LOG_ ## level, virtio_vdpa_logtype, \
+		"VIRTIO VDPA %s(): " fmt "\n", __func__, ##args)
+
+TAILQ_HEAD(virtio_vdpa_privs, virtio_vdpa_priv) virtio_priv_list =
+						  TAILQ_HEAD_INITIALIZER(virtio_priv_list);
+static pthread_mutex_t priv_list_lock = PTHREAD_MUTEX_INITIALIZER;
+
+static struct virtio_vdpa_priv *
+virtio_vdpa_find_priv_resource_by_vdev(const struct rte_vdpa_device *vdev)
+{
+	struct virtio_vdpa_priv *priv;
+	int found = 0;
+
+	pthread_mutex_lock(&priv_list_lock);
+	TAILQ_FOREACH(priv, &virtio_priv_list, next) {
+		if (vdev == priv->vdev) {
+			found = 1;
+			break;
+		}
+	}
+	pthread_mutex_unlock(&priv_list_lock);
+	if (!found) {
+		DRV_LOG(ERR, "Invalid vDPA device: %s", vdev->device->name);
+		rte_errno = ENODEV;
+		return NULL;
+	}
+	return priv;
+}
+
+static int
+virtio_vdpa_vqs_max_get(struct rte_vdpa_device *vdev, uint32_t *queue_num)
+{
+	struct virtio_vdpa_priv *priv =
+		virtio_vdpa_find_priv_resource_by_vdev(vdev);
+
+	if (priv == NULL) {
+		DRV_LOG(ERR, "Invalid vDPA device: %s", vdev->device->name);
+		return -ENODEV;
+	}
+	// TO_DO:for net, should use max_virtqueue_pairs? may be is same
+	*queue_num = priv->hw_nr_virtqs;
+	DRV_LOG(DEBUG, "Vid %d queue num is %d", priv->vid, *queue_num);
+	return 0;
+}
+
+static int
+virtio_vdpa_features_get(struct rte_vdpa_device *vdev, uint64_t *features)
+{
+	struct virtio_vdpa_priv *priv =
+		virtio_vdpa_find_priv_resource_by_vdev(vdev);
+
+	if (priv == NULL) {
+		DRV_LOG(ERR, "Invalid vDPA device: %s", vdev->device->name);
+		return -ENODEV;
+	}
+
+	virtio_dev_get_features(priv->vpdev,features);
+	*features |= (1ULL << VHOST_USER_F_PROTOCOL_FEATURES);
+
+	return 0;
+}
+
+static int
+virtio_vdpa_protocol_features_get(struct rte_vdpa_device *vdev,
+		uint64_t *features)
+{
+	struct virtio_vdpa_priv *priv =
+		virtio_vdpa_find_priv_resource_by_vdev(vdev);
+
+	if (priv == NULL) {
+		DRV_LOG(ERR, "Invalid vDPA device: %s", vdev->device->name);
+		return -ENODEV;
+	}
+	*features = VIRTIO_VDPA_PROTOCOL_FEATURES;
+	return 0;
+}
+
+static uint64_t
+virtio_vdpa_hva_to_gpa(int vid, uint64_t hva)
+{
+	struct rte_vhost_memory *mem = NULL;
+	struct rte_vhost_mem_region *reg;
+	uint32_t i;
+	uint64_t gpa = 0;
+
+	if (rte_vhost_get_mem_table(vid, &mem) < 0) {
+		if (mem)
+			free(mem);
+		DRV_LOG(ERR, "Virtio dev %d get mem table fail", vid);
+		return 0;
+	}
+
+	for (i = 0; i < mem->nregions; i++) {
+		reg = &mem->regions[i];
+
+		if (hva >= reg->host_user_addr &&
+				hva < reg->host_user_addr + reg->size) {
+			gpa = hva - reg->host_user_addr + reg->guest_phys_addr;
+			break;
+		}
+	}
+
+	free(mem);
+	return gpa;
+}
+
+static int
+virtio_vdpa_vfio_intr_enable(struct virtio_vdpa_priv *priv)
```

Comment:
No.

This code belongs in the low level virtio pci driver at virtio_pci_dev level.